draco/asp/hard.lp

```diff
@@ -1,24 +1,97 @@
 % ====== Hard constraints ======
 
-% @hard(text_mark_without_text_channel) Text mark requires text encoding.
-violation(text_mark_without_text_channel) :-
-    attribute((mark,type),M,text),
-    not attribute(mark_channel,M,text).
+% @hard(enc_type_valid) Primitive type has to support data type.
+violation(enc_type_valid) :-
+    attribute(field_type,N,(string;boolean)),
+    attribute(scale_type,M,C,T),
+    attribute(channel_field,M,C,N),
+    T != ordinal.
 
-% @hard(text_channel_without_text_mark) Text channel requires text mark.
-violation(text_channel_without_text_mark) :-
-    attribute(mark_channel,M,text),
-    not attribute((mark,type),M,text).
+% @hard(bin_q_o) Can only bin quantitative or ordinal.
```

Comment:
Hmm, I'm not ready for binning yet. How binning relates to scales is something we need to think about a bit more. Let's remove binning constraints from the pull request for now.

Reply:
removed.